src/openrct2/paint/map_element/virtual_floor.c

```diff
@@ -0,0 +1,296 @@
+        //  * Ghost objects, which are displayed as lit squares
+    rct_map_element * mapElement = map_get_first_element_at(session->MapPosition.x >> 5, session->MapPosition.y >> 5);
+    do
+    {
+        sint32 elementType = map_element_get_type(mapElement);
+
+        if (elementType == MAP_ELEMENT_TYPE_SURFACE)
+        {
+            if (virtualFloorClipHeight < mapElement->clearance_height)
+            {
+                weAreBelowGround    = true;
+            }
+            else if (virtualFloorClipHeight < mapElement->base_height + 2 &&
+                 mapElement->properties.surface.slope != 0)
+            {
+                weAreBelowGround    = true;
+                weAreOccupied       = true;
+            }
+            continue;
+        }
+
+        if (virtualFloorClipHeight >= mapElement->clearance_height ||
+            virtualFloorClipHeight < mapElement->base_height)
+        {
+            continue;
+        }
+
+        if (elementType == MAP_ELEMENT_TYPE_WALL ||
+            elementType == MAP_ELEMENT_TYPE_BANNER)
+        {
+            sint32 direction = map_element_get_direction(mapElement);
+            occupiedEdges   |= 1 << ((direction + get_current_rotation()) % 4);
+            continue;
+        }
+        
+        if (mapElement->flags & MAP_ELEMENT_FLAG_GHOST)
+        {
+            weAreLit = true;
+            continue;
+        }
+
+        weAreOccupied = true;
+    }
+    while (!map_element_is_last_for_tile(mapElement++));
+
+        // Try the four tiles next to us for the same parameters as above,
+        //  if our parameters differ we set an edge towards that tile
+    for (uint8 i = 0; i < 4; i++)
+    {
+        uint8 effectiveRotation = (4 + i - get_current_rotation()) % 4;
+        sint16 theirLocationX = session->MapPosition.x + offsets[effectiveRotation].x;
+        sint16 theirLocationY = session->MapPosition.y + offsets[effectiveRotation].y;
+
+        bool theyAreOccupied        = false;
+        bool theyAreBelowGround     = false;
+        bool theyAreLit             = false;
+        
+        if ((gMapSelectFlags & MAP_SELECT_FLAG_ENABLE) &&
+            theirLocationX >= gMapSelectPositionA.x &&
+            theirLocationY >= gMapSelectPositionA.y &&
+            theirLocationX <= gMapSelectPositionB.x &&
+            theirLocationY <= gMapSelectPositionB.y)
+        {
+            theyAreLit = true;
+        }
+        if (!theyAreLit)
+        {
+            if (gMapSelectFlags & MAP_SELECT_FLAG_ENABLE_CONSTRUCT)
+            {
+                for (tile = gMapSelectionTiles; tile->x != -1; tile++)
+                {
+                    if (theirLocationX == tile->x &&
+                        theirLocationY == tile->y)
+                    {
+                        theyAreLit = true;
+                    }
+                }
+            }
+        }
+
+        mapElement = map_get_first_element_at(theirLocationX >> 5, theirLocationY >> 5);
+        do
+        {
```

Comment:
This whole loop and the condition above looks pretty much the same as the `we` variant of it. Any chance you could refactor them into their own function and just call it for different tiles?